docs/content/en/docs/Design/layout.md

```diff
@@ -0,0 +1,608 @@
+layout is assigned. The default layout for scalars is to repeat the scalar in
+every slot of a single ciphertext. The default layout for tensors is a row-major
+layout into as many ciphertexts as are needed to fit the tensor.
+
+Then layouts are propagated forward through the IR. For each op, a default
+kernel is chosen, and if the layouts of the operands are already set and agree,
+the result layout is inferred according to the kernel.
+
+If the layouts are not compatible with the default kernel, a `convert_layout` op
+is inserted to force compatibility. If one or more operands has a layout that is
+not set (which can happen if the operand is a cleartext value known to the
+server), then a compatible layout is chosen and an `assign_layout` op is
+inserted to persist this information for later passes.
+
+Because `layout-propagation` may have inserted some redundant conversions,
+sequences of `assign_layout` followed by `convert_layout` are folded together
+into combined `assign_layout` ops.
+
+**FIXME: how are things chosen when the tensor sizes are not a power of two?**
+
+### `layout-optimization`
+
+The [`layout-optimization`](/docs/passes/#-layout-optimization) pass has two
+main goals: to choose better kernels for ops, and to try to eliminate
+`convert_layout` ops. It does this by running a backward pass through the IR. If
+it encounters an op that is followed by a `convert_layout` op, it attempts to
+hoist the `convert_layout` through the op to its arguments.
+
+In doing this, it must consider:
+
+- Changing the kernel of the op, and the cost of implementing the kernel. E.g.,
+  a new kernel may be better for the new layout of the operands.
+- Whether the new layout of op results still need to be converted, and the new
+  cost of these conversions. E.g., if the op result has multiple uses, or the op
+  result had multiple layout conversions, only one of which is hoisted.
+- The new cost of operand layout conversions. E.g., if a layout conversion is
+  hoisted to one operand, it may require other operands to be converted to
+  remain compatible.
+
+In all of the above, the "cost" includes an estimate of the latency of a kernel,
+an estimate of the latency of a layout conversion, as well as the knowledge that
+some layout conversions may be free or cheaper because of their context in the
+IR.
+
+The cost of a kernel is hard-coded in a table.
+
+**FIXME: kernels are currently considered free, cf. #1888**
+
+The cost of a layout conversion is estimated by simulating what the
+`implement-shift-network` would do if it ran on a layout conversion. And
+`layout-optimization` includes analyses that allow it to determine a folded cost
+for layout conversions that occur after other layout conversions, as well as the
+free cost of layout conversions that occur at function arguments, after
+`assign_layout` ops, or separated from these by ops that do not modify a layout.
+
+**FIXME: extend this pass to also pick the min kernel for an op even if it
```

Comment:
@asraa I am considering trying to fix this FIXME and the above before merging this design doc. Thoughts? LMK if you had already started on either of these.